### PR TITLE
closed - weather

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Review by robske110 — COMMENTED** [review]
I think the PM codestyle is more of a lowercase bool. (You used Bool)
Nice work, will be intereating to see where it goes

**Comment by dktapps** [comment]
Looks interesting 🍿

File: src/pocketmine/level/WeatherManager.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	public function isWeatherEnabled(){
+		return $this->weatherEnabled;
+	}
+
+	public function toggleWeather(){
+		$this->setDuration(mt_rand(300, 6000));
+	}
+
+	public function sendWeatherToPlayers($players = null){
+		$players = $players !== null && is_array($players) ? $players : $this->getServer()->getOnlinePlayers();
```
Comment (dktapps) [inline src/pocketmine/level/WeatherManager.php]: are you sure about this? Weather is per world, just the same as time is.

Comment (madecode15) [inline src/pocketmine/level/WeatherManager.php]: Whoops, I'll fix that.

**Comment by madecode15** [comment]
@dktapps 

Which would you prefer?
`$player->sendWeather($id);`
or
`$this->getLevel()->getWeatherManager()->sendWeatherToPlayer($player, $id);`

I think the first option is cleaner to code with.

**Comment by robske110** [comment]
:one:
Why?
Because it makes no sense to need the weather manager for individual player weathers.

File: src/pocketmine/level/WeatherManager.php [hunk 2]
```diff
@@ -112,8 +114,19 @@ public function sendWeatherToPlayers($players = null){
...
+		$pk = new AddEntityPacket();
+		$pk->type = 93;
+		$pk->eid = Entity::$entityCount++;
+		$pk->x = $pos->z;
+		$pk->y = $pos->y;
+		$pk->z = $pos->z;
+		$pk->yaw = $yaw;
+        $pk->pitch = $pitch;
+		$pk->metadata = is_array($metadata) ? $metadata : []; //Not sure what defualt values should be.
+		foreach($this->getServer()->getOnlinePlayers() as $p){
```
Comment (jasonwynn10) [inline src/pocketmine/level/WeatherManager.php]: probably should be only for the specific level and not for everyone online

Comment (madecode15) [inline src/pocketmine/level/WeatherManager.php]: @jasonwynn10 Did you see the commit made a few seconds after that one?

Comment (jasonwynn10) [inline src/pocketmine/level/WeatherManager.php]: no, no I didn't. Sorry!

**Comment by madecode15** [comment]
@JackNoordhuis 

Correct.
> There is however no information on how random these events are in single player.

This will be implemented.
> The wiki states that a thunderstorm can occur randomly at any time in the overworld and that rain has a small chance to worsen into a thunderstorm.

File: src/pocketmine/level/WeatherManager.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	public function strikeLighting(Vector3 $pos, $yaw, $pitch, $metadata = null){
+		$pk = new AddEntityPacket();
+		$pk->type = 93;
+		$pk->eid = Entity::$entityCount++;
+		$pk->x = $pos->z;
+		$pk->y = $pos->y;
+		$pk->z = $pos->z;
+		$pk->yaw = $yaw;
+        $pk->pitch = $pitch;
```
Comment (robske110) [inline src/pocketmine/level/WeatherManager.php]: small indent fail

File: src/pocketmine/Player.php [hunk 4]
```diff
@@ -3026,6 +3029,18 @@ public function sendMessage($message){
 		*/
 	}
 
+	/**
+	 * Send a weather mode to aplayer.
```
Comment (robske110) [inline src/pocketmine/Player.php]: A space is missing

**Review by robske110 — COMMENTED** [review]
It's really starting to take shape, good work.
I think weatherEnabled is a bit confusing, but i can't come up with sth better either.

**Comment by dktapps** [comment]
Actually, I'm pretty sure I've read somewhere that random weather can last anywhere from 30 seconds to 5 minutes. I could be wrong, and I can't cite a source because I don't remember where I read it. Probably Minecraft Wiki.

File: src/pocketmine/level/WeatherManager.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+				}
+				$this->setWeather(self::RAIN);
+				break;
+			case self::RAIN:
+				$this->setWeather(self::NORMAL);
+				break;
+		}
+	}
+
+	public function sendWeatherToPlayers($players = null){
```
Comment (dktapps) [inline src/pocketmine/level/WeatherManager.php]: Empty array as default would be better I think, along with an `array` typehint.

File: src/pocketmine/Player.php [hunk 6]
```diff
@@ -811,6 +812,8 @@ protected function doFirstSpawn(){
 		$pk->z = $pos->z;
 		$this->dataPacket($pk);
 
+		$this->sendWeather($this->getLevel()->getWeatherManager()->getWeather());
```
Comment (dktapps) [inline src/pocketmine/Player.php]: There are fields in StartGamePacket for this. If you can use those instead that would be better.

**Comment by dktapps** [comment]
By the way, you might want to fix your gitconfig. Your commits aren't associated with your account.

File: src/pocketmine/level/WeatherManager.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		foreach($players as $p){
+			$p->dataPacket($pk);
+		}
+	}
+
+	public function strikeLighting(Vector3 $pos, $yaw, $pitch, array $metadata = []){
+		$pk = new AddEntityPacket();
+		$pk->type = 93;
+		$pk->eid = Entity::$entityCount++;
+		$pk->x = $pos->z;
```
Comment (jojoe77777) [inline src/pocketmine/level/WeatherManager.php]: ;P

Comment (madecode15) [inline src/pocketmine/level/WeatherManager.php]: hm?

Comment (jojoe77777) [inline src/pocketmine/level/WeatherManager.php]: X is not Z

**Comment by dktapps** [comment]
Please bear in mind: 
https://github.com/pmmp/PocketMine-MP/blob/master/CONTRIBUTING.md#contributing-code

> Avoid using GitHub Web Editor. The web editor lacks most useful GIT features and should only be used for very minor changes. It is immediately clear if the web editor has been used, and if so the PR is more likely to be rejected. If you want to make serious contributions, please learn how to use GIT version control.

**Comment by madecode15** [comment]
@dktapps Oh ok, I''ll close this for now then.

**Comment by HimbeersaftLP** [comment]
> Please bear in mind:
https://github.com/pmmp/PocketMine-MP/blob/master/CONTRIBUTING.md#contributing-code
Avoid using GitHub Web Editor. The web editor lacks most useful GIT features and should only be used for very minor changes. It is immediately clear if the web editor has been used, and if so the PR is more likely to be rejected. If you want to make serious contributions, please learn how to use GIT version control.

So a contribution with the GitHub web editor isn't a real contribution?! That's like saying "Your E-Mail isn't a real E-Mail because it was sent from a Webmail client" or "This movie isn't a real movie because you watched it on Netflix".

Code = Code it doesn't matter where it came from

**Comment by dktapps** [comment]
um...no. It means don't use the web editor because it's crap, and code written with the web editor is unlikely to pass QA, not to mention that lots of spam commits is annoying.

**Pull request closed without merging**